keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -91,242 +92,324 @@ with the unreasonable development overhead of a webhook.
+   
+   Both extensions can only be set in `v1` CRDs on creation, but updated in `v1beta1` if preexisting.
    
-   If that comparison fails, the request is rejected as 400 "Bad Request".
+2. to use **strict deep-equal** comparison of those fields marked as immutable during
+   update validation in the request pipeline, with the only exception outlined below in the detailed semantics of the different values of the two vendor extensions.
    
 We create another KEP to define custom normalization steps for CRs done during 
 deserialization from etcd and when receiving a request (just after pruning and defaulting).
 
-### OpenAPI extension `x-kubernetes-immutable`
+### Recursive Scope
+
+If the `x-kubernetes-mutability` vendor extension is set (to `Immutable`, `AddOnly`, `RemoveOnly`), `x-kubernetes-mutability: Immutable` is implicitly applied recursively to all subschemas.
+
+The `x-kubernetes-key-mutability` vendor extension does not apply recursively.
+
+### Semantics of `x-kubernetes-mutability`
+
+#### Immutable list, arbitrary list type
+
+The whole list is immutable in the strict sense of `reflect.DeepEqual`.
+
+#### Immutable map / object, arbitrary map type
+
+The whole map is immutable in the strict sense of `reflect.DeepEqual`.
+
+#### Mutable object, immutable fields
+
+The value of a field is immutable, and it cannot be added (Immutable, RemoveOnly) or removed (Immutable, AddOnly).
 
-The `x-kubernetes-immutable` vendor extension is set in the spec of fields,
-arrays and objects. It recursively applies to the whole subtree:
+*Example 1 - Immutable*:
+
+```yaml
+properties:
+  foo:
+    type: string
+    x-kubernetes-mutability: Immutable | AddOnly | RemoveOnly
+```
+
+Allowed for AddOnly, disallowed otherwise:
+- `{}` → `{foo:"a"}`
+
+Allowed for RemoveOnly, disallowed otherwise:
+- `{foo:"a"}` → `{}`
+
+Disallowed:
+- `{foo:"a"}` →	`{foo:"b"}`
+
+#### Mutable array, immutable items, list type list
+
+The items with their respective index in the array are immutable. Hence, appending and removal at the end of the array are allowed, the change of an item or the change of the order are not.
+
+*Example 2*:
 
 ```yaml
 properties:
   foo:
-    type: array
     items:
-      x-kubernetes-immutable: true
-      type: string
+      type: object
+      properties: ...
+      x-kubernetes-mutability: Immutable
 ```
 
-means that the `foo` array itself is mutable, but the items are not. This means that
-items can be deleted and added, but at each existing index the string values are
-immutable.
+Allowed:
+- `{}` → `{foo:...}`
+- `{foo:...}` → `{}`
+- `{foo:[a]}` → `{foo:[a,b]}` (value of key 0 is unchanged)
+- `{foo:[a]}` →	`{foo:[]}`
+
+Disallowed:
+- `{foo:[a]}` → `{foo:[b]}`
+- `{foo:[a]}` → `{foo:[b,a]}` (value of key 0 is changed)
+
+#### Mutable array, immutable items, list type map
+
+The key-value pairs in the array are immutable, set set of keys is not. Hence, addition and removal of key-values pairs
```

Comment:
The key-value pairs in the array are immutable, ~set~ set of keys is not. Hence, addition and removal of key-values pairs is allowed